### PR TITLE
Add a guide for `nnx.bridge`

### PR DESCRIPTION
**Pull request by IvyZX**
Todo for future guides:

* When we have a guide for NNX's JAX-style transform, cross-reference it here.
* When we have a standalone Linen-to-NNX (refactored from the current Haiku/Linen/NNX guide), cross reference it here.

### CODE REVIEW

File: docs/nnx/bridge_guide.md [hunk 1]
```diff
@@ -0,0 +1,544 @@
+# Use NNX along with Linen
+
+This guide is for existing Flax users who want to make their codebase a mixture of Linen and NNX modules, which is made possible by the `nnx.bridge` API. 
+
+This will be helpful if you:
+
+* Want to migrate your codebase to NNX gradually, one module at a time;
+* Have external dependency that already moved to NNX but you haven't, or is still in Linen while you've moved to NNX.
+
+We hope this allows you to move and try out NNX at your own pace, and leverage the best of both worlds. We will also talk about how to resolve the coveats of interoping the two APIs, on a few aspects that they are fundamentally different.
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:10]:
```suggestion
We hope this allows you to move and try out NNX at your own pace, and leverage the best of both worlds. We will also talk about how to resolve the caveats of interoperating the two APIs, on a few aspects that they are fundamentally different.
```

Comment (IvyZX) [reply docs/nnx/bridge_guide.md]: Done!

**Review by 8bitmp3 — COMMENTED** [review]
First observations, LMKWYT

File: docs/nnx/bridge_guide.md [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+from jax import numpy as jnp
+from jax.experimental import mesh_utils
+from typing import *
+```
+
+## Submodule is all you need
+
+A Flax model is always a tree of modules - either old Linen modules (`flax.linen.Module`, usually written as `nn.Module`) or NNX modules (`nnx.Module`). 
+
+An `nnx.bridge` wrapper glues the two types together, in both ways:
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:36]:
```suggestion
The `nnx.bridge` wrapper glues the two types together, in both ways:
```

File: docs/nnx/bridge_guide.md [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+  out_dim: int
+  w_init: Callable[..., Any] = nn.initializers.lecun_normal()
+  @nn.compact
+  def __call__(self, x):
+    # Linen might need the input shape to create the weight!
+    w = self.param('w', self.w_init, (x.shape[-1], self.out_dim))
+    return x @ w
+
+x = jax.random.normal(jax.random.key(42), (4, 32))
+model = bridge.ToNNX(LinenDot(64), rngs=nnx.Rngs(0))  # => `model = LinenDot(64)` in Linen
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:72]: It would be interesting to  do an `nnx.display` here after construction to visualize the uninitialized structure, and in a new cell call the rest and visualize the module again after its initialized.

File: docs/nnx/bridge_guide.md [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+assert isinstance(model.dot.params['w'].value, jax.Array)
+```
+
+### NNX -> Linen
+
+To convert an NNX module to Linen, you should forward your creation arguments to `bridge.ToLinen` and let it handle the actual creation process. 
+
+We don't let you create an NNX module of your own because new variables are created eagerly as soon as a NNX module is created. Instead, `bridge.ToLinen` will handle it for you.
+
+An NNX module instance initializes all its variables eagerly when it is created, which consumes memory and compute. On the other hand, Linen modules are stateless, and the typical `init` and `apply` process involves multiple creation of them. To bypass this issue, you should send your creation arguments (instead of a created NNX module) to `bridge.to_linen`, and let it handle the rest.
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:128]: There is a bit of repetition here.

File: docs/nnx/bridge_guide.md [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+
+```python
+class NNXDot(nnx.Module):
+  def __init__(self, in_dim: int, out_dim: int, rngs: nnx.Rngs):
+    self.w = nnx.Param(nnx.initializers.lecun_normal()(rngs.params(), (in_dim, out_dim)))
+  def __call__(self, x: jax.Array):
+    return x @ self.w
+
+x = jax.random.normal(jax.random.key(42), (4, 32))
+model = bridge.ToLinen(NNXDot, args=(32, 64))  # <- Pass in the arguments, not an actual module
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:139]: We should use `to_linen` as its both more ergonomic and does the `FrozenDict` conversion automatically. Most users will copy the first example they see.

File: docs/nnx/bridge_guide.md [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
+```
+
+    <class 'flax.nnx.nnx.graph.NodeDef'>
+
+
+You can use a shortcut `bridge.to_linen` to avoid explicitly grouping args and kwargs separately. You only need to use the underlying `bridge.ToLinen()` in some rare cases.
+
+
+```python
+bridge.to_linen(NNXDot, 32, out_dim=64) == bridge.ToLinen(NNXDot, args=(32,), kwargs={'out_dim': 64})
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:169]: I feel that `to_linen` should be the main API. Lets not show 2 ways of doing this.

File: docs/nnx/bridge_guide.md [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+
+```python
+# This new field stores the static data that defines the underlying `NNXDot`
+print(type(variables['nnx']['graphdef']))    # => `nnx.graph.NodeDef`
+```
+
+    <class 'flax.nnx.nnx.graph.NodeDef'>
+
+
+You can use a shortcut `bridge.to_linen` to avoid explicitly grouping args and kwargs separately. You only need to use the underlying `bridge.ToLinen()` in some rare cases.
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:165]: Maybe we can also forward the other arguments such that you can always use `to_linen`?

File: docs/nnx/bridge_guide.md [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+```
+
+    (32, 64) (1, 64) (4, 64)
+
+
+## Handling RNG keys
+
+All Flax modules, Linen or NNX, automatically handle the RNG keys for variable creation and random layers like dropouts. However, the specific logics of RNG key splitting are different, so you cannot generate the same params between Linen and NNX modules, even if you pass in same keys.
+
+Another difference is that NNX modules are stateful, so they can track and update the RNG keys within themselves. This means a bit difference in both `ToNNX` and `ToLinen`:
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:219]:
```suggestion
Another difference is that NNX modules are stateful, so they can track and update the RNG keys within themselves. This means RNG handling is bit different in both `ToNNX` and `ToLinen`:
```

Comment (cgarciae) [inline docs/nnx/bridge_guide.md:219]:
```
This means a bit difference in both `ToNNX` and `ToLinen`
```
This part is a bit unclear.

File: docs/nnx/bridge_guide.md [hunk 9]
```diff
@@ -0,0 +1,544 @@
...
+model = bridge.ToNNX(nn.Dropout(rate=0.5, deterministic=False), rngs=nnx.Rngs(dropout=0))
+bridge.lazy_init(model, x)       # We don't really need this b/c no extra params were created here,
+                                 # but it's a good practice to always add this line.
+y1, y2 = model(x), model(x)
+assert not jnp.allclose(y1, y2)  # Two runs yield different outputs!
+
+# Reset the dropout RNG seed, so that next model run will be the same as the first.
+nnx.reseed(model, dropout=0)
+assert jnp.allclose(y1, model(x))
+```
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:237]: Love this example!

**Comment by cgarciae** [comment]
One general thing I'd add is the use of `nnx.display` on most cells so users can visualize the structure, it makes it a bit more intuitive to follow e.g:

Screenshot 2024-09-06 at 11 01 56

File: docs/nnx/bridge_guide.md [hunk 10]
```diff
@@ -0,0 +1,544 @@
...
+
+# The NNX RNG state was stored inside `variables`
+print('The RNG key in state:', variables['RngKey']['rngs']['dropout']['key'].value)
+print('How many times we have created new RNG keys from it:', variables['RngCount']['rngs']['dropout']['count'].value)
+
+# NNX style: Must set `RngCount` as mutable and update the variables after every `apply`
+y1, updates = model.apply(variables, x, mutable=['RngCount'])
+variables |= updates
+y2, updates = model.apply(variables, x, mutable=['RngCount'])
+assert not jnp.allclose(y1, y2)  # Every call yields different output!
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:265]:
```suggestion
print('Key split count:', variables['RngCount']['rngs']['dropout']['count'].value)

# NNX style: Must set `RngCount` as mutable and update the variables after every `apply`
y1, updates = model.apply(variables, x, mutable=['RngCount'])
variables |= updates
y2, updates = model.apply(variables, x, mutable=['RngCount'])
variables |= updates

print('Key split count:', variables['RngCount']['rngs']['dropout']['count'].value)
assert not jnp.allclose(y1, y2)  # Every call yields different output!
```

File: docs/nnx/bridge_guide.md [hunk 11]
```diff
@@ -0,0 +1,544 @@
...
+
+# Linen style: Just pass different RNG keys for every `apply()` call.
+y3 = model.apply(variables, x, rngs={'dropout': jax.random.key(1)})
+y4 = model.apply(variables, x, rngs={'dropout': jax.random.key(2)})
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:269]: could be interesting to do the opposite and use the same key to show that this overrides the keys and yields the same output

File: docs/nnx/bridge_guide.md [hunk 12]
```diff
@@ -0,0 +1,544 @@
...
+
+
+### NNX to Linen
+
+If you define custom NNX variable types, you should register their names with `nnx.register_variable_name_type_pair` so that they go to the desired collections.
+
+
+```python
+class Count(nnx.Variable): pass
+nnx.register_variable_name_type_pair('Count', Count, overwrite=True)
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:365]: `Count` would be the automatic name anyway, maybe we can show registering a Linen-style name (no capital, plural)?
```suggestion
nnx.register_variable_name_type_pair('counts', Count, overwrite=True)
```

File: docs/nnx/bridge_guide.md [hunk 13]
```diff
@@ -0,0 +1,544 @@
...
+  @nn.compact
+  def __call__(self, x):
+    w = self.param('w', nn.with_partitioning(nn.initializers.lecun_normal(), ('in', 'out')), 
+                   (x.shape[-1], self.out_dim))
+    return x @ w
+
+@jax.jit
+def create_sharded_nnx_module(x):
+  model = bridge.lazy_init(bridge.ToNNX(LinenDotWithPartitioning(64), rngs=nnx.Rngs(0)), x)
+  static, state = nnx.split(model)
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:421]: We started using `graphdef` instead of `static` as its a bit more descriptive.

File: docs/nnx/bridge_guide.md [hunk 14]
```diff
@@ -0,0 +1,544 @@
...
+  sharded_state = nnx.with_sharding_constraint(state, nnx.get_partition_spec(state))
+  return static, sharded_state
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:423]: Consider using `nnx.jit` here so you don't have to use `merge` later. 
```suggestion
@nnx.jit
def create_sharded_nnx_module(x):
  model = bridge.lazy_init(bridge.ToNNX(LinenDotWithPartitioning(64), rngs=nnx.Rngs(0)), x)
  state = nnx.state(model)
  sharded_state = nnx.with_sharding_constraint(state, nnx.get_partition_spec(state))
  nnx.update(model, sharded_state)
  return model
```

File: docs/nnx/bridge_guide.md [hunk 15]
```diff
@@ -0,0 +1,544 @@
...
+
+
+print(f'We have {len(jax.devices())} fake JAX devices now to partition this model...')
+mesh = jax.sharding.Mesh(devices=mesh_utils.create_device_mesh((2, 4)), axis_names=('in', 'out'))
+x = jax.random.normal(jax.random.key(42), (4, 32))
+with mesh:
+  model = nnx.merge(*create_sharded_nnx_module(x))
```
Comment (cgarciae) [inline docs/nnx/bridge_guide.md:430]: (If the previous is implemented)
```suggestion
  model = create_sharded_nnx_module(x)
```

**Review by cgarciae — APPROVED** [review]

File: docs/nnx/bridge_guide.md [hunk 16]
```diff
@@ -0,0 +1,544 @@
...
+* Want to migrate your codebase to NNX gradually, one module at a time;
+* Have external dependency that already moved to NNX but you haven't, or is still in Linen while you've moved to NNX.
+
+We hope this allows you to move and try out NNX at your own pace, and leverage the best of both worlds. We will also talk about how to resolve the coveats of interoping the two APIs, on a few aspects that they are fundamentally different.
+
+**Note**: 
+
+This guide is about glueing Linen and NNX modules. To migrate an existing Linen module to NNX, check out the [Migrate from Linen to NNX](https://flax.readthedocs.io/en/latest/nnx/haiku_linen_vs_nnx.html) guide. 
+
+And all built-in Linen layers should have equivalent NNX versions! Check out the list of [Built-in NNX layers](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/index.html).
```
[the remark this answers is not in the document]
Comment (IvyZX) [reply docs/nnx/bridge_guide.md:16]: Done.

File: docs/nnx/bridge_guide.md [hunk 17]
```diff
@@ -0,0 +1,544 @@
+# Use NNX along with Linen
```
[the remark this answers is not in the document]
Comment (IvyZX) [reply docs/nnx/bridge_guide.md:1]: Done.

File: docs/nnx/bridge_guide.md (hunk 6 above)
Comment (IvyZX) [reply docs/nnx/bridge_guide.md:169]: Done. Explained a use case where you must use `ToLinen`.

File: docs/nnx/bridge_guide.md (hunk 5 above)
Comment (IvyZX) [reply docs/nnx/bridge_guide.md:139]: Done.

File: docs/nnx/bridge_guide.md (hunk 4 above)
Comment (IvyZX) [reply docs/nnx/bridge_guide.md:128]: Done.

File: docs/nnx/bridge_guide.md (hunk 3 above)
Comment (IvyZX) [reply docs/nnx/bridge_guide.md:72]: Good idea! Done.

File: docs/nnx/bridge_guide.md [hunk 18]
```diff
@@ -0,0 +1,544 @@
...
+
+This guide is for existing Flax users who want to make their codebase a mixture of Linen and NNX modules, which is made possible by the `nnx.bridge` API.
```
[the remark this answers is not in the document]
Comment (IvyZX) [reply docs/nnx/bridge_guide.md:3]: Done.

File: docs/nnx/bridge_guide.md (hunk 8 above)
Comment (IvyZX) [reply docs/nnx/bridge_guide.md:219]: Removed.

File: docs/nnx/bridge_guide.md (hunk 10 above)
Comment (IvyZX) [reply docs/nnx/bridge_guide.md:265]: Sounds great! Done

File: docs/nnx/bridge_guide.md (hunk 11 above)
Comment (IvyZX) [reply docs/nnx/bridge_guide.md:269]: Good idea! done.

File: docs/nnx/bridge_guide.md (hunk 14 above)
Comment (IvyZX) [reply docs/nnx/bridge_guide.md:423]: Done. I should get more used to it!

File: docs/nnx/bridge_guide.md (hunk 12 above)
Comment (IvyZX) [reply docs/nnx/bridge_guide.md:365]: Done.

File: docs/nnx/bridge_guide.md (hunk 15 above)
Comment (IvyZX) [reply docs/nnx/bridge_guide.md:430]: Done.

File: docs/nnx/bridge_guide.md [hunk 19]
```diff
@@ -0,0 +1,669 @@
+# Use Flax NNX along with Flax Linen
+
+This guide is for existing Flax users who want to make their codebase a mixture of Flax Linen and Flax NNX `Module`s, which is made possible thanks to the `flax.nnx.bridge` API.
```
Comment (ayaka14732) [inline docs/nnx/bridge_guide.md:3]: There are redundant trailing spaces on several lines.

```suggestion
This guide is for existing Flax users who want to make their codebase a mixture of Flax Linen and Flax NNX `Module`s, which is made possible thanks to the `flax.nnx.bridge` API.
```

Comment (IvyZX) [reply docs/nnx/bridge_guide.md:3]: Thanks Ayaka! It seems that our internal & external code checkers doesn't flag out trailing spaces in notebook markdowns so I got lazy and didn't fix them... Will keep an eye out on the next PR!
